amm/contracts/router_v2/pool.rs

```diff
@@ -0,0 +1,147 @@
+use amm_helpers::{ensure, math::casted_mul};
+use ink::primitives::AccountId;
+use traits::{MathError, Pair as PairTrait, RouterV2Error, StablePool as StablePoolTrait};
+
+const PAIR_TRADING_FEE_DENOM: u128 = 1000;
+
+pub use crate::pair::*;
+pub use crate::stable_pool::*;
+
+#[derive(scale::Decode, scale::Encode)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+)]
+pub enum Pool {
+    Pair(Pair),
+    StablePool(StablePool),
+}
+
+impl Pool {
+    pub fn get_amount_in(
+        &self,
+        token_in: AccountId,
+        token_out: AccountId,
+        amount_out: u128,
+    ) -> Result<u128, RouterV2Error> {
+        match self {
+            Pool::Pair(pool) => {
+                let (reserve_in, reserve_out) = pool.get_reserves(token_in, token_out);
```

Comment:
We copy the v2 logic here because the `Pair` contract does not have methods for swap estimations. Notice that we still have to do a cross-contract call to get the reserves. I'm not sure why it's designed this way but that's how the original Router contract works:
https://github.com/Cardinal-Cryptography/common-amm/blob/392345129a9f017fc0ae7c90af00c5243bd28e77/amm/contracts/router/lib.rs#L244-L247
https://github.com/Cardinal-Cryptography/common-amm/blob/392345129a9f017fc0ae7c90af00c5243bd28e77/amm/contracts/router/lib.rs#L670-L704

Reply:
A bit unlucky, but guess we don't have a better choice then